### ISSUE TITLE
Liquid Exception: incompatible character encodings: UTF-8 and ASCII-8BIT

### ISSUE THREAD

**Issue by MaxLaumeister** [jekyll/jekyll#8133]
As per the issue template:

  - I updated to the latest Jekyll (or) if on GitHub Pages to the latest `github-pages`
  - I ran `jekyll doctor` to check my configuration
  - I read the contributing document at https://jekyllrb.com/docs/contributing/

## My Environment

| Software              | Version(s) |
| ------------------ | ---------- |
| Operating System |  Windows 10  |
| `jekyll`         | 4.0.0 |
| `github-pages`   | No |

---

## Expected Behaviour

The page should be generated, containing the content of all other pages on the site.

## Current Behavior

I get the following error when building:

```
Liquid Exception: incompatible character encodings: UTF-8 and ASCII-8BIT in search/index.html
```

Full Output:

```
> bundle exec jekyll serve --trace

Configuration file: C:/Users/...redacted.../_config.yml
            Source: C:/Users/...redacted...
       Destination: C:/Users/...redacted.../_site
 Incremental build: disabled. Enable with --incremental
      Generating...
  Liquid Exception: incompatible character encodings: UTF-8 and ASCII-8BIT in search/index.html
Traceback (most recent call last):
        42: from C:/Ruby26-x64/bin/jekyll:23:in `<main>'
        41: from C:/Ruby26-x64/bin/jekyll:23:in `load'
        40: from C:/Ruby26-x64/lib/ruby/gems/2.6.0/gems/jekyll-4.0.0/exe/jekyll:15:in `<top (required)>'
        39: from C:/Ruby26-x64/lib/ruby/gems/2.6.0/gems/mercenary-0.3.6/lib/mercenary.rb:19:in `program'
        38: from C:/Ruby26-x64/lib/ruby/gems/2.6.0/gems/mercenary-0.3.6/lib/mercenary/program.rb:42:in `go'
        37: from C:/Ruby26-x64/lib/ruby/gems/2.6.0/gems/mercenary-0.3.6/lib/mercenary/command.rb:220:in `execute'
        36: from C:/Ruby26-x64/lib/ruby/gems/2.6.0/gems/mercenary-0.3.6/lib/mercenary/command.rb:220:in `each'
        35: from C:/Ruby26-x64/lib/ruby/gems/2.6.0/gems/mercenary-0.3.6/lib/mercenary/command.rb:220:in `block in execute'
        34: from C:/Ruby26-x64/lib/ruby/gems/2.6.0/gems/jekyll-4.0.0/lib/jekyll/commands/serve.rb:86:in `block (2 levels) in init_with_program'
        33: from C:/Ruby26-x64/lib/ruby/gems/2.6.0/gems/jekyll-4.0.0/lib/jekyll/command.rb:89:in `process_with_graceful_fail'
        32: from C:/Ruby26-x64/lib/ruby/gems/2.6.0/gems/jekyll-4.0.0/lib/jekyll/command.rb:89:in `each'
        31: from C:/Ruby26-x64/lib/ruby/gems/2.6.0/gems/jekyll-4.0.0/lib/jekyll/command.rb:89:in `block in process_with_graceful_fail'
        30: from C:/Ruby26-x64/lib/ruby/gems/2.6.0/gems/jekyll-4.0.0/lib/jekyll/commands/build.rb:36:in `process'
        29: from C:/Ruby26-x64/lib/ruby/gems/2.6.0/gems/jekyll-4.0.0/lib/jekyll/commands/build.rb:65:in `build'
        28: from C:/Ruby26-x64/lib/ruby/gems/2.6.0/gems/jekyll-4.0.0/lib/jekyll/command.rb:28:in `process_site'
        27: from C:/Ruby26-x64/lib/ruby/gems/2.6.0/gems/jekyll-4.0.0/lib/jekyll/site.rb:76:in `process'
        26: from C:/Ruby26-x64/lib/ruby/gems/2.6.0/gems/jekyll-4.0.0/lib/jekyll/site.rb:202:in `render'
        25: from C:/Ruby26-x64/lib/ruby/gems/2.6.0/gems/jekyll-4.0.0/lib/jekyll/site.rb:515:in `render_pages'
        24: from C:/Ruby26-x64/lib/ruby/gems/2.6.0/gems/jekyll-4.0.0/lib/jekyll/site.rb:515:in `each'
        23: from C:/Ruby26-x64/lib/ruby/gems/2.6.0/gems/jekyll-4.0.0/lib/jekyll/site.rb:516:in `block in render_pages'
        22: from C:/Ruby26-x64/lib/ruby/gems/2.6.0/gems/jekyll-4.0.0/lib/jekyll/site.rb:523:in `render_regenerated'
        21: from C:/Ruby26-x64/lib/ruby/gems/2.6.0/gems/jekyll-4.0.0/lib/jekyll/renderer.rb:63:in `run'
        20: from C:/Ruby26-x64/lib/ruby/gems/2.6.0/gems/jekyll-4.0.0/lib/jekyll/renderer.rb:80:in `render_document'
        19: from C:/Ruby26-x64/lib/ruby/gems/2.6.0/gems/jekyll-4.0.0/lib/jekyll/renderer.rb:127:in `render_liquid'
        18: from C:/Ruby26-x64/lib/ruby/gems/2.6.0/gems/jekyll-4.0.0/lib/jekyll/liquid_renderer/file.rb:32:in `render!'
        17: from C:/Ruby26-x64/lib/ruby/gems/2.6.0/gems/jekyll-4.0.0/lib/jekyll/liquid_renderer/file.rb:60:in `measure_time'
        16: from C:/Ruby26-x64/lib/ruby/gems/2.6.0/gems/jekyll-4.0.0/lib/jekyll/liquid_renderer/file.rb:33:in `block in render!'
        15: from C:/Ruby26-x64/lib/ruby/gems/2.6.0/gems/jekyll-4.0.0/lib/jekyll/liquid_renderer/file.rb:53:in `measure_bytes'
        14: from C:/Ruby26-x64/lib/ruby/gems/2.6.0/gems/jekyll-4.0.0/lib/jekyll/liquid_renderer/file.rb:34:in `block (2 levels) in render!'
        13: from C:/Ruby26-x64/lib/ruby/gems/2.6.0/gems/jekyll-4.0.0/lib/jekyll/liquid_renderer/file.rb:49:in `measure_counts'
        12: from C:/Ruby26-x64/lib/ruby/gems/2.6.0/gems/jekyll-4.0.0/lib/jekyll/liquid_renderer/file.rb:35:in `block (3 levels) in render!'
        11: from C:/Ruby26-x64/lib/ruby/gems/2.6.0/gems/liquid-4.0.3/lib/liquid/template.rb:220:in `render!'
        10: from C:/Ruby26-x64/lib/ruby/gems/2.6.0/gems/liquid-4.0.3/lib/liquid/template.rb:207:in `render'
         9: from C:/Ruby26-x64/lib/ruby/gems/2.6.0/gems/liquid-4.0.3/lib/liquid/template.rb:242:in `with_profiling'
         8: from C:/Ruby26-x64/lib/ruby/gems/2.6.0/gems/liquid-4.0.3/lib/liquid/template.rb:208:in `block in render'
         7: from C:/Ruby26-x64/lib/ruby/gems/2.6.0/gems/liquid-4.0.3/lib/liquid/block_body.rb:82:in `render'
         6: from C:/Ruby26-x64/lib/ruby/gems/2.6.0/gems/liquid-4.0.3/lib/liquid/block_body.rb:103:in `render_node_to_output'
         5: from C:/Ruby26-x64/lib/ruby/gems/2.6.0/gems/liquid-4.0.3/lib/liquid/tags/for.rb:79:in `render'
         4: from C:/Ruby26-x64/lib/ruby/gems/2.6.0/gems/liquid-4.0.3/lib/liquid/tags/for.rb:150:in `render_segment'
         3: from C:/Ruby26-x64/lib/ruby/gems/2.6.0/gems/liquid-4.0.3/lib/liquid/context.rb:123:in `stack'
         2: from C:/Ruby26-x64/lib/ruby/gems/2.6.0/gems/liquid-4.0.3/lib/liquid/tags/for.rb:158:in `block in render_segment'
         1: from C:/Ruby26-x64/lib/ruby/gems/2.6.0/gems/liquid-4.0.3/lib/liquid/tags/for.rb:158:in `each'
C:/Ruby26-x64/lib/ruby/gems/2.6.0/gems/liquid-4.0.3/lib/liquid/tags/for.rb:160:in `block (2 levels) in render_segment': incompatible character encodings: UTF-8 and ASCII-8BIT (Encoding::CompatibilityError)
```

## Config File

I've redacted some variables that definitely have nothing to do with this error. Here's the rest:

```
# Netlify-specific settings
include: [
    _headers,
    _redirects
]

collections:
    example:
        output: true
        permalink: /:collection/:name/

defaults:
  - scope:
      path: ""
      type: "example"
    values:
      layout: "example2"
      description: "example description text"

# Build settings
markdown: kramdown
future: false

exclude: [
    vendor,
    Gemfile,
    Gemfile.lock,
    README.md
]

plugins: []
```

## Code Sample

/search/index.html

```
---
---
 {% for post in site.pages %}
   {{ post.content }}
 {% endfor %}
```

The intent behind this was to further process this output into json and use it to build a search index.

Unlike #2629, the only plugin I have installed is `liquify_filter.rb`, for which the only code is this:

```
module Jekyll
  module LiquifyFilter
    def liquify(input)
      Liquid::Template.parse(input).render(@context)
    end
  end
end

Liquid::Template.register_filter(Jekyll::LiquifyFilter)
```

I also found another related issue from 2015: #3969

**Comment by ashmaroli** [comment]
The focus of the error is:
```
~/gems/liquid-4.0.3/lib/liquid/tags/for.rb:160:in `block (2 levels) in render_segment':
incompatible character encodings: UTF-8 and ASCII-8BIT (Encoding::CompatibilityError)
```
which means something wrong with the rendered segment of the Liquid `for` block (in `search/index.html`).

Since you're iterating `{{ page.content }}` through `site.pages`, I'd say one of your *pages* are the source of your issue.
It won't be possible for us to pin-point which page exactly without having access to the source repository.

**Comment by MaxLaumeister** [comment]
Got it. I'll see if I can narrow it down to a specific page that's causing the issue.

**Comment by MaxLaumeister** [comment]
I did a manual binary search to try to figure out which specific page was causing the issue.

It ended up being a `.css.map` file in my assets folder. Apparently it was included in Jekyll's `site.pages` which I was looping through.

I excluded my entire assets folder from the list of pages to process, and that fixed the issue.

Thanks for the tip, ashmaroli. Maybe this thread will be useful to someone in the future.
